pkg/ignore/dockerignore.go

```diff
@@ -0,0 +1,242 @@
+
+var (
+	_            Ignorer = &defaultIgnorer{}
+	excludedDirs         = sets.NewString()
+	includesDirs         = sets.NewString()
+	directory    string
+)
+
+// dockerIgnorer processes .dockerignore and use the extracted patterns for ignoring files
+type dockerIgnorer struct {
+	patterns *sets.String
+}
+
+// NewOrDefault builds and returns the new or default Ignorer interface
+func NewOrDefault(dir string) (Ignorer, error) {
+	directory = dir
+	//Always include the root directory
+	includesDirs.Insert(dir)
+
+	//Ensure the re-init it every time
+	patterns := sets.NewString(DefaultPatterns...)
+
+	ignoreFile := filepath.Join(dir, DockerIgnoreFile)
+	_, err := os.Stat(ignoreFile)
+
+	if os.IsNotExist(err) {
+		return &defaultIgnorer{
+			&patterns,
+		}, nil
+	}
+
+	fr, err := os.Open(ignoreFile)
+
+	switch {
+	case os.IsNotExist(err):
+		return &defaultIgnorer{
+			&patterns,
+		}, nil
+	case err != nil:
+		return &defaultIgnorer{
+			&patterns,
+		}, err
+	}
+
+	scanner := bufio.NewScanner(fr)
+
+	utf8bom := []byte{0xEF, 0xBB, 0xBF}
+	lineNo := 0
+
+	for scanner.Scan() {
+		scannedBytes := scanner.Bytes()
+
+		if lineNo == 0 {
+			scannedBytes = bytes.TrimPrefix(scannedBytes, utf8bom)
+		}
+
+		pattern := string(scannedBytes)
+		pattern = strings.TrimSpace(pattern)
+		lineNo++
+
+		// Ignore comments
+		if strings.HasPrefix("#", pattern) {
+			continue
+		}
+
+		// Ignore Empty lines
+		if pattern == "" {
+			continue
+		}
+
+		// Normalize - clean and slash
+		invert := pattern[0] == '!'
+		if invert {
+			pattern = strings.TrimSpace(pattern[1:])
+		}
+
+		if len(pattern) > 0 {
+			pattern = filepath.Clean(pattern)
+			pattern = filepath.ToSlash(pattern)
+			if len(pattern) > 1 && pattern[0] == '/' {
+				pattern = pattern[1:]
+			}
+		}
+
+		// add invert to the cleaned pattern
+		if invert {
+			pattern = "!" + pattern
+		}
+
+		patterns.Insert(pattern)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return &defaultIgnorer{
+			&patterns,
+		}, fmt.Errorf("error reading file %s : %w", ignoreFile, err)
+	}
+
+	defer fr.Close()
+
+	return &dockerIgnorer{
+		&patterns,
+	}, err
+}
+
+// Ignore implements Ignorer, for  dockerignore cases
+func (id *dockerIgnorer) Ignore(path string, fi os.FileInfo) Ignoreable {
+	bundleFile := &bundleFile{
+		name:     fi.Name(),
+		path:     path,
+		rootDir:  directory,
+		isDir:    fi.IsDir(),
+		patterns: id.patterns,
+	}
+
+	ignorable := &Ignoreable{}
+
+	if directory != path {
+		if !checkIfParentIncluded(path, fi) {
+			ignorable.Ignored = bundleFile.ignore()
+			if fi.IsDir() {
+				ignorable.Type = Transitive
+			} else {
+				ignorable.Type = Current
+			}
+
+		}
+	}
+
+	return *ignorable
+}
+
+// Patterns implements Ignorer, returns the pattern associated with the ignorer
+func (id *dockerIgnorer) Patterns() *sets.String {
+	return id.patterns
+}
+
+func checkIfParentIncluded(path string, fi os.FileInfo) bool {
+	//if parent is included then dont check them children, include them implicitly
```

Comment:
This seems like it should be a method comment?

Reply:
done